lib/Driver/ToolChains.cpp

```diff
@@ -847,6 +848,74 @@ toolchains::Darwin::constructInvocation(const InterpretJobAction &job,
   return II;
 }
 
+static StringRef
+getDarwinLibraryNameSuffixForTriple(const llvm::Triple &triple) {
+  switch(getDarwinPlatformKind(triple)) {
+    case DarwinPlatformKind::MacOS:
+      return "osx";
+    case DarwinPlatformKind::IPhoneOS:
+      return "ios";
+    case DarwinPlatformKind::IPhoneOSSimulator:
+      return "iossim";
+    case DarwinPlatformKind::TvOS:
+      return "tvos";
+    case DarwinPlatformKind::TvOSSimulator:
+      return "tvossim";
+    case DarwinPlatformKind::WatchOS:
+      return "watchos";
+    case DarwinPlatformKind::WatchOSSimulator:
+      return "watchossim";
+  }
+  llvm_unreachable("Unsupported Darwin platform");
+}
+
+static void AddLinkRuntimeLib(const ArgList &Args, ArgStringList &Arguments,
+                              StringRef DarwinLibName, bool AddRPath,
+                              const ToolChain &TC) {
+  SmallString<128> Dir;
+  getRuntimeLibraryPath(Dir, Args, TC);
+  // Remove platform name.
+  llvm::sys::path::remove_filename(Dir);
+  llvm::sys::path::append(Dir, "clang", "lib", "darwin");
+  SmallString<128> P(Dir);
+  llvm::sys::path::append(P, DarwinLibName);
+  Arguments.push_back(Args.MakeArgString(P));
+
+  // Adding the rpaths might negatively interact when other rpaths are involved,
```

Comment:
This should be in the doc comment for this function and everything that calls it, *and* the top-level call site.